### PR TITLE
CUDA Lite kernels for parallel memcpy and memset

### PR DESCRIPTION
**Pull request by bornaehsani**
Added memcpy and memset cuda lite runtime kernels
User can now call a grid of tile groups to perform parallel memset and memcpy on the device DRAM with the desired size and value, instead of having to do it through slow PCI-E.
Kernel code can be incorporated inside your larger programs.

Related to PR # 435 in bsg_f1.

### CODE REVIEW

File: software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#include "bsg_set_tile_x_y.h"
+
+#define BSG_TILE_GROUP_X_DIM bsg_tiles_X
+#define BSG_TILE_GROUP_Y_DIM bsg_tiles_Y
+#include "bsg_tile_group_barrier.h"
+INIT_TILE_GROUP_BARRIER(r_barrier, c_barrier, 0, bsg_tiles_X-1, 0, bsg_tiles_Y-1);
+
+
+
+int  __attribute__ ((noinline)) kernel_memcpy(int *dst, int *src, int size) {
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c]: *src and size should be annotated with `const`

File: software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+	int __bsg_tile_group_id = (__bsg_tile_group_id_y * __bsg_grid_dim_x + __bsg_tile_group_id_x);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c]: Does every invocation have to do this computation? Why not have it computed in the launch function?

Comment (drichmond) [reply software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c]: (This is probably a separate PR, not to be resolve here)

File: software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	int __bsg_tile_group_dim = (bsg_tiles_X * bsg_tiles_Y);
+        int block_size_x = size / (__bsg_grid_dim_y * __bsg_grid_dim_x);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c]: Whitespace issue

File: software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+#define BSG_TILE_GROUP_Y_DIM bsg_tiles_Y
+#include "bsg_tile_group_barrier.h"
+INIT_TILE_GROUP_BARRIER(r_barrier, c_barrier, 0, bsg_tiles_X-1, 0, bsg_tiles_Y-1);
+
+
+
+int  __attribute__ ((noinline)) kernel_memcpy(int *dst, int *src, int size) {
+
+	int __bsg_tile_group_id = (__bsg_tile_group_id_y * __bsg_grid_dim_x + __bsg_tile_group_id_x);
+	int __bsg_tile_group_dim = (bsg_tiles_X * bsg_tiles_Y);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c]: Did you mean _size? A dimension is two variables (X and Y) in my mind.

File: software/spmd/bsg_cuda_lite_runtime/memset/kernel_memset.c [hunk 5]
```diff
@@ -1,16 +1,30 @@
...
 #define BSG_TILE_GROUP_X_DIM bsg_tiles_X
 #define BSG_TILE_GROUP_Y_DIM bsg_tiles_Y
 #include "bsg_tile_group_barrier.h"
 INIT_TILE_GROUP_BARRIER(r_barrier, c_barrier, 0, bsg_tiles_X-1, 0, bsg_tiles_Y-1);
 
 
-int  __attribute__ ((noinline)) kernel_memset() {
-  bsg_tile_group_barrier(&r_barrier, &c_barrier);
+
+int  __attribute__ ((noinline)) kernel_memset(int *A, int val, int size) {
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/memset/kernel_memset.c]: I would rename *A to *dst, and add const specifiers to val and size

File: software/spmd/bsg_cuda_lite_runtime/memset/kernel_memset.c [hunk 6]
```diff
@@ -1,16 +1,30 @@
...
+
+	int __bsg_tile_group_id = (__bsg_tile_group_id_y * __bsg_grid_dim_x + __bsg_tile_group_id_x);
+	int __bsg_tile_group_dim = (bsg_tiles_X * bsg_tiles_Y);
+        int block_size_x = size / (__bsg_grid_dim_y * __bsg_grid_dim_x);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/memset/kernel_memset.c]: Whitespace issue

File: software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c (hunk 2 above)
Comment (bornaehsani) [reply software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c]: Good point! I actually made a PR on bsg_f1 (#425) to add the __bsg_tile_group_id symbol to the binary, and have it set by the host. It went stale at some point, but I'll rebase it and make a PR again, once that is merged in I can go ahead and remove these invocations, (I intentionally named the variable the same as symbol so once that PR is merged in this throws and error so I know to fix it.)

File: software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c (hunk 4 above)
Comment (bornaehsani) [reply software/spmd/bsg_cuda_lite_runtime/memcpy/kernel_memcpy.c]: renamed. Thanks!

**Comment by bornaehsani** [comment]
@drichmond Thanks for the comments, I addressed them, I think we're good to go. Let me know!

**Pull request merged**